### PR TITLE
Canvas keys: move to global and make improve focus trapping

### LINKED ISSUE #1070: Keyboard shortcuts do not work when multiple elements selected
## Feature Description
When multiple elements are selected, keyboard shortcut do not work. So arrow keys to move and delete, should move or remove all elements elements.

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### LINKED ISSUE #1781: Delete key requires focusing on canvas before removing selected elements
## Bug Description

Delete key seems to work only if canvas has focus first, resulting it deleting elements sometimes not working as expected.

## Expected Behaviour

Delete key should work on a global level, with taking into consideration active selection or input focus -- in these cases, the Delete key should have default behavior (see the [`nativeCopyPasteExpected.js` ](https://github.com/google/web-stories-wp/blob/master/assets/src/edit-story/utils/nativeCopyPasteExpected.js) util which could perhaps be renamed to `nativeEventHandlingExpected.js` or sth similar to use it for this occasion, too)

## Steps to Reproduce

1. Select an element (or more)
2. Click on one of the library tabs, e.g. shapes
3. Hit `Delete`.
4. See nothing happening.

## Screenshots

Example on FF where clicking "Delete" key tries to go to the previous browser history entry:
delete

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### LINKED ISSUE #1808: Pressing 'Enter' on a text element is not entering in edit mode sometimes
## Bug Description

Pressing <kbd>Enter</kbd> in a selected text element sometimes not put that text element in edit mode.

## Expected Behaviour

Text elements selected in display mode should always go to edit mode when pressing <kbd>Enter</kbd>.

## Steps to Reproduce

Scenario 1:
1. Select a text element (display mode) 
2. Press <kbd>Enter</kbd>
3. Text element will go to Edit mode
4. Click outside the text element to leave the edit mode (without unselecting the text element)
5. Text element will go to Display Mode
6. Press <kbd>Enter</kbd> again

Scenario 2:
1. Add a text element
2. Press <kbd>Enter</kbd>

## Screenshots

Untitled

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### LINKED ISSUE #2417: Change how global shortcuts work and make canvas shortcuts global
## Feature Description

Dependent bugs: #1781, #1070, #1808.

The canvas is the central part of the editor. As such most of keyboard shortcuts have to apply to the canvas, unless they conflict with another component currently focused.

The goal is to make canvas shortcuts global, but to change the rules to resolve conflicts.

## Alternatives Considered

We tried to make canvas area always focused. But this kind of focus trapping is an anathema in a Web application and breaks a lot of accessibility features.

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Acceptance Criteria / QA Instructions

1. Add a new text element. Click "Delete". The element should be deleted.
2. Add a new text element. Click "left" or "right". The element should move.
1. Select an element on the canvas, click on the "Text" tab in the library. Click delete. The element should be deleted.
2. Select a text element on the canvas, click on "B" (bold) button in the design panel. Click delete. The element should be deleted.
3. Add a new page. Add an element. Click on the current page in the carousel. Click "left" arrow key. The first page should be selected. The element should't be moved. Click "delete". The page should be remove.
4. Select a text element. Click on a color selector. Observer color picker is open. Click "delete". The element should not be deleted.

## Implementation Brief

1. Make focus trapping should take into account single vs multi-selection. If a single element is selected, this elements frame should be focused. Otherwise, the canvas itself should be focused.
2. All of the canvas key handlers will be moved to global handlers.
3. In most of places where only up/down or left/right were handled, the relevant components should handle both pairs and just no-op them if no implementation is possible. This is done so that key handling is not fragmented: e.g. up/down works and left/right does something else.
4. Global shortcuts are to be enabled on most buttons/checkboxes/etc. I.e. clicking Delete on a button is should trigger a global shortcut.
5. The global shortcuts should be separated by the dialog boundaries. The idea is that shortcuts should not make changes that the user cannot even see, b/c a dialog is blocking their view.

### PR DESCRIPTION
**Pull request by dvoytenko**
## Summary

Changes how canvas focus and keyboard are done.

## Relevant Technical Choices

The key points:
 1. Focus trapper is slightly smarter now: when only one element is selected, it will focus the frame of that element. Otherwise the behavior is unchanged and the canvas itself is selected.
 2. Key handlers have been added to the page carousel for testing and feature completeness. See testing instructions below.
 3. All of the canvas key handlers have been moved to global handlers. Ht #2189.
 4. In most of places if only up/down or left/right were handled, now the relevant components handle both pairs and just no-op them if no implementation is possible. This is done so that key handling is not fragmented: e.g. up/down works and left/right does something else.
 5. Shortcuts are enabled over most of buttons/checkboxes/etc. I.e. clicking Delete on a button is not ignored by Mousetrap. This is a mousetrap feature and it's done by selectively adding "mousetrap" class to the affected DOM elements.
 6. Important! The global shortcuts are by default disallowed from crossing over dialog boundaries. E.g. a global shortcut may or may not work inside a page grid or a color picker. This is configurable. But idea is that shortcuts should not make changes that the user cannot even see, b/c a dialog is blocking their view.
 7. Minor: outlines are added to toggle buttons. This helps testing.

## To-do

- [x] Tests

## User-facing changes

See relevant bugs reports. Additionally, keyboard shortcuts have been implemented in the page carousel.

## Testing Instructions

See relevant bugs reports.

Additionally, for the page carousel, testing instructions are:
 1. Left/right keys should navigate between the carousel pages.
 2. Command+left/right should reorder the carousel pages.
 3. Delete key should delete the selected carousel page.
 4. Command+D should duplicate the selected page.
 5. Esc key should exit the carousel and go back to the canvas.

---

Fixes #2417
Fixes #1781 
Fixes #1070 
Fixes #1808

### CODE REVIEW

File: assets/src/edit-story/components/canvas/canvasProvider.js [hunk 1]
```diff
@@ -48,6 +48,7 @@ function CanvasProvider({ children }) {
     setEditingElementWithState,
     setEditingElementWithoutState,
     clearEditing,
+    getNodeForElement,
```
Comment (barklund) [inline assets/src/edit-story/components/canvas/canvasProvider.js:51]: I **hate** how big this component is! I think it's our worst one in the entire codebase. Buuuuut that's outside the scope of this ticket.

File: assets/src/edit-story/components/canvas/carousel/useCarouselKeys.js [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+      setCurrentPage,
+    })
+  );
+
+  const pageIds = useMemo(() => pages.map(({ id }) => id), [pages]);
+
+  useEscapeToBlurEffect(ref);
+
+  // Navigate left and right.
+  useKeyDownEffect(
```
Comment (barklund) [inline assets/src/edit-story/components/canvas/carousel/useCarouselKeys.js:63]: Don't we need to `stopPropagation()` here? Or is that default in the key handler?

File: assets/src/edit-story/components/form/radioGroup.js [hunk 3]
```diff
@@ -53,7 +53,8 @@ const Name = styled.span`
   line-height: 24px;
 `;
 
-const Radio = styled.input`
+// Class should contain "mousetrap" to enable keyboard shortcuts on buttons.
```
Comment (barklund) [inline assets/src/edit-story/components/form/radioGroup.js]:
```suggestion
// Class should contain "mousetrap" to enable keyboard shortcuts on inputs.
```

File: assets/src/edit-story/components/form/switch.js [hunk 4]
```diff
@@ -50,9 +50,11 @@ const SwitchContainer = styled.div`
   border: none;
 `;
 
+// Class should contain "mousetrap" to enable keyboard shortcuts on buttons.
```
Comment (barklund) [inline assets/src/edit-story/components/form/switch.js]:
```suggestion
// Class should contain "mousetrap" to enable keyboard shortcuts on inputs.
```

File: assets/src/edit-story/components/form/toggle.js [hunk 5]
```diff
@@ -26,7 +26,11 @@ import { Fragment } from 'react';
  */
 import WithTooltip from '../tooltip';
 
-const CheckBoxInput = styled.input.attrs({ type: 'checkbox' })`
+// Class should contain "mousetrap" to enable keyboard shortcuts on buttons.
```
Comment (barklund) [inline assets/src/edit-story/components/form/toggle.js]:
```suggestion
// Class should contain "mousetrap" to enable keyboard shortcuts on inputs.
```

File: assets/src/edit-story/components/form/toggleButton.js [hunk 6]
```diff
@@ -24,9 +24,14 @@ import { rgba } from 'polished';
 /**
  * Internal dependencies
  */
+import { KEYBOARD_USER_SELECTOR } from '../../utils/keyboardOnlyOutline';
 import MULTIPLE_VALUE from './multipleValue';
 
-const CheckBoxInput = styled.input.attrs({ type: 'checkbox' })`
+// Class should contain "mousetrap" to enable keyboard shortcuts on buttons.
```
Comment (barklund) [inline assets/src/edit-story/components/form/toggleButton.js]:
```suggestion
// Class should contain "mousetrap" to enable keyboard shortcuts on inputs.
```

File: assets/src/edit-story/components/keyboard/index.js [hunk 7]
```diff
@@ -295,6 +300,13 @@ function isEditableTarget({ tagName, isContentEditable, type, readOnly }) {
   return false;
 }
 
+function crossesDialogBoundary(target, keyTarget) {
+  // Check if somewhere between `keyTarget` and `target` there's a
+  // dialog boundary.
```
Comment (barklund) [inline assets/src/edit-story/components/keyboard/index.js:310]: Uh, nice catch here.

**Review by barklund — COMMENTED** [review]
This works really well. I like it! A few comments, but otherwise awesome!

**Comment by dvoytenko** [comment]
Tests have been added.

File: assets/src/edit-story/components/canvas/canvasProvider.js (hunk 1 above)
Comment (dvoytenko) [reply assets/src/edit-story/components/canvas/canvasProvider.js:51]: Agree. Definitely unnecessarily large.

File: assets/src/edit-story/components/canvas/carousel/useCarouselKeys.js (hunk 2 above)
Comment (dvoytenko) [reply assets/src/edit-story/components/canvas/carousel/useCarouselKeys.js:63]: Yes, `stopPropagation()` is always default with Mousetrap.

**Comment by dvoytenko** [comment]
@barklund I merged your suggestions, but now it asks for your CLA for some reason.

**Comment by barklund** [comment]
@googlebot I consent.

**Comment by barklund** [comment]
@googlebot I fixed it.

**Comment by dvoytenko** [comment]
I tried bunch of things but couldn't get CLA to cooperate. I ended up squashing the commits to fix it :(

**Review by barklund — APPROVED** [review]
This is awesome!

However we need to make sure to note all impacted areas for a full QA on those areas - making sure that keyboard and mouse interactions still work as expected.

File: assets/src/edit-story/components/canvas/karma/carousel.karma.js [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+    await fixture.events.keyboard.press('right');
+    expect(await getCurrentPageId()).toEqual(page4);
+  });
+
+  it('should reorder the page with keys', async () => {
+    await fixture.events.mouse.clickOn(getThumbnail(1), 5, 5);
+    expect(await getCurrentPageId()).toEqual(page2);
+
+    // Order left.
+    await fixture.events.keyboard.shortcut('mod+left');
```
Comment (miina) [inline assets/src/edit-story/components/canvas/karma/carousel.karma.js:160]: Note: currently in the compact view, it's not possible to reorder pages from carousel using mouse, however, now it's possible via keyboard. I guess that's fine there's a difference?

Comment (dvoytenko) [reply assets/src/edit-story/components/canvas/karma/carousel.karma.js:160]: Probably not very meaningful. I filed https://github.com/google/web-stories-wp/issues/2575 to clarify.

**Pull request merged**